src/shims/fs.rs

```diff
@@ -139,6 +150,25 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             } else {
                 this.handle_not_found()
             }
+        } else if cmd == this.eval_libc_i32("F_DUPFD")? || cmd == this.eval_libc_i32("F_DUPFD_CLOEXEC")? {
+            // Note that we always assume the FD_CLOEXEC flag is set for every open file, in part
+            // because exec() isn't supported. The F_DUPFD and F_DUPFD_CLOEXEC commands only
+            // differ in whether the FD_CLOEXEC flag is pre-set on the new file descriptor,
+            // thus they can share the same implementation here.
+            let arg_op = arg_op
+                .ok_or_else(|| err_unsup_format!("fcntl with command F_DUPFD or F_DUPFD_CLOEXEC requires a third argument"))?;
+            let arg = this.read_scalar(arg_op)?.to_i32()?;
+            let fh = &mut this.machine.file_handler;
+            let (file_result, writable) = match fh.handles.get(&fd) {
+                Some(original) => (original.file.try_clone(), original.writable),
+                None => return this.handle_not_found(),
+            };
+            let fd_result = file_result.map(|duplicated| {
+                let new_fd = std::cmp::max(fh.next_fd(), arg);
```

Comment:
What is the point of this "max" here? Shouldn't it be impossible for `arg` to be bigger than the next fresh FD?

Reply:
The docs for `F_DUPFD` say:

>Find the lowest numbered available file descriptor greater than or equal to arg and make it be a copy of fd.

We need to consider both what the next available fd is, and this third argument, which could be anything.

Reply:
`arg` is a valid FD. All valid FDs are smaller than the "next fresh" FD. So at least in the current impl, it's impossible for the `max` to matter.

Reply:
Maybe the `low` approach just makes no sense, and we should have a "fresh_fd" function that determines the lowest unused FD by walking the HashMap, and use that?

Reply:
> Maybe the `low` approach just makes no sense, and we should have a "fresh_fd" function that determines the lowest unused FD by walking the HashMap, and use that?

https://github.com/rust-lang/miri/pull/1156#discussion_r373782188

Reply:
I don't know, in my reading of the manpage, I don't see that arg has to be a valid file descriptor. The errors section only says that it will raise EINVAL if arg is negative or too large, there's nothing about EBADF if arg isn't an open fd.

I like the idea of walking the HashMap, that might be the clearest approach.

Reply:
> I like the idea of walking the HashMap, that might be the clearest approach.
Well I suppose its not like anyone is expecting to open a million files, I think that works

Reply:
> I don't see that arg has to be a valid file descriptor

Ah, I misread the code.

But, still, the code here seems wrong to me -- the code has an invariant that `low` is the largest file descriptor ever assigned (+/- 1), and the `max` computation here can result in that being violated.

> I like the idea of walking the HashMap, that might be the clearest approach.

If we make it a BTreeMap, it shouldn't even be inefficient to search for the smallest free key in some range.